Closes #274: Token Rollout to all workflows and parameter files

ahmadabdalla opened the pull request:
## Change 

Closes #295 

1) Applying the 'Default Tokens' to all workflows so they can now start consuming them:

```yaml
  ARM_SUBSCRIPTION_ID: '${{ secrets.ARM_SUBSCRIPTION_ID }}'
  ARM_MGMTGROUP_ID: '${{ secrets.ARM_MGMTGROUP_ID }}'
  ARM_TENANT_ID: '${{ secrets.ARM_TENANT_ID }}'
  DEPLOYMENT_SP_ID: '${{ secrets.DEPLOYMENT_SP_ID }}'
  PLATFORM_KEYVAULT: '${{ secrets.PLATFORM_KEYVAULT }}'
```

2) Global Module Tests
  - 2 Test cases to validate the use of Tokens (subscriptionId, tenantId)
  - Wanted to add the 'principalId' check but that needs more testing as there are many variations to this

3) Applied the following tokens (on all the resources)
 - `<<deploymentSpId>>` is now the default for all role assignments.
 - `<<subscriptionId>>` is now the default for subscription
 - `<<managementGroupId>>` is now the default for management group deployments, enabled for `parentId` in one of the parameter files
 - `<<tenantId>>` is now the default for any mentioning of the tenant that is part of this environment
 
## Other changes
- readme files that contain hardcoded values have now been tokenized. **Not** because we want to tokenize them, but we should use a generic reference to a subscription Id, and the token version achieves that. 
- one of the constructs was using the 'artifacts-rg', now swapped with <<resourceGroupName>>

## Out of scope of this change
- moving resource names to use the `<<namePrefix>>` token in the resource name. This needs more testing for every resource which will take a lot of time. Guidance would be for every module being worked on; we swap the name with the token (if we want to).
- test cases for tokens: This is hard to construct for custom tokens given it is unknown to the test framework. Needs more thinking to get it right (if we want to enable it).
- Dependency Pipeline: Key Vault Access Policies now reference the `<<tenantId>>` token

## Testing

[![Sql: Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.sql.servers.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.sql.servers.yml)
[![Network: Expressroutecircuits](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.expressroutecircuits.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.expressroutecircuits.yml)
[![Network: Networkwatchers](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.networkwatchers.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.networkwatchers.yml)
[![Network: Privateendpoints](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.privateendpoints.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.privateendpoints.yml)
[![AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.analysisservices.servers.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.analysisservices.servers.yml)
[![ContainerRegistry: Registries](https://github.com/Azure/ResourceModules/actions/workflows/ms.containerregistry.registries.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.containerregistry.registries.yml)
[![Insights: Privatelinkscopes](https://github.com/Azure/ResourceModules/actions/workflows/ms.insights.privatelinkscopes.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.insights.privatelinkscopes.yml)
[![Insights: Actiongroups](https://github.com/Azure/ResourceModules/actions/workflows/ms.insights.actiongroups.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.insights.actiongroups.yml)
[![RecoveryServices: Vaults](https://github.com/Azure/ResourceModules/actions/workflows/ms.recoveryservices.vaults.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.recoveryservices.vaults.yml)
[![Storage: Storage Account](https://github.com/Azure/ResourceModules/actions/workflows/ms.storage.storageaccounts.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.storage.storageaccounts.yml)
[![ServiceBus: Namespaces](https://github.com/Azure/ResourceModules/actions/workflows/ms.servicebus.namespaces.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.servicebus.namespaces.yml)
[![Network: Publicipaddresses](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.publicipaddresses.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.publicipaddresses.yml)
[![Network: Networksecuritygroups](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.networksecuritygroups.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.networksecuritygroups.yml)
[![Network: Applicationsecuritygroups](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.applicationsecuritygroups.yml/badge.svg)](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.applicationsecuritygroups.yml)
[![KeyVault: Vaults](https://github.com/Azure/ResourceModules/actions/workflows/ms.keyvault.vaults.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.keyvault.vaults.yml)
[![Compute: Diskencryptionsets](https://github.com/Azure/ResourceModules/actions/workflows/ms.compute.diskencryptionsets.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.compute.diskencryptionsets.yml)
[![Authorization: Policyassignments](https://github.com/Azure/ResourceModules/actions/workflows/ms.authorization.policyassignments.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.authorization.policyassignments.yml)
[![Authorization: policyDefinitions](https://github.com/Azure/ResourceModules/actions/workflows/ms.authorization.policydefinitions.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.authorization.policydefinitions.yml)
[![ManagedServices: Registrationdefinitions](https://github.com/Azure/ResourceModules/actions/workflows/ms.managedservices.registrationdefinitions.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/ms.managedservices.registrationdefinitions.yml)

[![.Platform: Dependencies](https://github.com/Azure/ResourceModules/actions/workflows/platform.dependencies.yml/badge.svg?branch=users%2Fahmad%2FtokenRollout)](https://github.com/Azure/ResourceModules/actions/workflows/platform.dependencies.yml)

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [ ] I did format my code

eriqua reviewed (commented):
A couple of questions after a first overview:
- Is the custom keyvault mandatory or still an option? How to test modules using e.g. remoteTenantId without the custom keyvault implemented?
- We currently have a few child restructuring PRs automation account, sqlmi, galleries, private dns, managed clusters. Should we wait for them to be merged first?
- Will the global test fail if values in the settings.json are not tokenized?

ahmadabdalla commented:
@eriqua:
- In our repository, the `remoteTenantId` is required because we use it to test and release the module that needs it. If the repo is forked and the person using it wanted to deploy and test this module, we have the 'Test-ModuleLocally' as a method to help and swap tokens. If they want to perform complete e2e testing in their fork like what we do, then we can give guidance in the documentation for that specific module (i.e. README.md for the automationAccount) on the tokens it uses and what they need to do to have the same experience for testing and validation. Thoughts?
- The changes only impact the workflow (+ env variables), parameter files (swap with tokens), and readme (swap with tokens). I don't see conflicts happening unless there is an impact to the same lines as what this PR is doing. (automationAccountName --> name, shouldn't impact this PR)
- No. The goal for global tests is to target the 'default tokens' only, even that is hard to achieve from what I have been trialing but would love more insights when we introduce more tests. The reason is that there may/will be different variations of what is inside the token, and even if we convert the parameter file to object, the placement of these tokens in the object can be at different locations, or even in sub-properties, which would require to build a complex script to manage, and still be error prone.

eriqua commented:
@ahmadabdalla 
1) I find all this a bit challenging to explain clearly to people just starting using this repo. The 'Test-ModuleLocally' is a utility and should not be the first option under certain circumstances i.e. depending on which module is used. Providing too many options could be challenging both to maintain and to describe in a step-by-step guidance. We should decouple clearly the module library from the built-in testing pipeline (test-validate-deploy-publish) that we are providing. We should then avoid to mention anything related to the latter in the readme.md, which is intended as a documentation for the module itself. I don't mean any destructive feedback. Tokenization is a great work (I'm a huge supporter) primarily intended to ease leveraging this repo.  We should just avoid to add complexity instead. That's my opinion at least and I'm curious about other's one. @MrMCake @MariusStorhaug @simonbMS @rahalan what's your take on that?
2) Parameter files are normally refactored quite extensively to include child resources. We may introduce additional values to be tokenized there.
3) Makes sense and I agree 👍

ahmadabdalla commented:
@eriqua your point is completely valid so no destructive feedback taken. The main problem space we want to address is to decouple the repo content from our own configurations/settings while still being able to release changes, accept contributions with minimal onboarding, keeping user experience at top of mind and not be constrained with the pipeline technologies. We have to draw the line somewhere and make trade offs. The use case for remote tokens is very narrow (for now) but it is a use case. Do we provide guidance on how to handle this use case (I.e prerequisites to have a key vault) via the wiki, or we pivot to another design? I'd say 90% of the use cases would be handled by default tokens, 9% with local tokens, 1% remote tokens.
I can definitely wait for a restructure and see the conflicts, no issues here

ahmadabdalla commented:
Discussion moved internally

MrMCake commented:
Just to add my 5 cents to the 3 raised points

1. We should make ourselves aware that we're talking about changes in the parameter files - not the template files. Hence I'd strongly suggest NOT to touch the module ReadMe files. As @eriqua stated, they should not have a reference to our testing whatsoever. Using `Test-ModuleLocally` as a function to test locally seems more a necessity to me rather than something 'optionally'. If we want to have tokens, the parameter files won't run without a wrapper - it is what is is. However, as the example in the wiki is quite big, I' suggest to trim it down a bit and not show every single optional option you can hand to the script. Rather have that taken care of by the readMe. In my opinion, the main example in the wiki should who the tokens & maybe the switches to enable/disable e.g. validation tests, deployment tests & static tests. Nothing more.
2. We merged quite a lot of PRs by now - let's see what comes up. To be on the safe site @ahmadabdalla I'd suggest to merge main already back into your PR. Just in case
3. I was already wondering about that. However I understand that testing this aint easy. I think enforcing e.g. the token for critical information like the subscriptionId is a good start

ahmadabdalla commented:
Ok latest update:
- Removed <<remoteTenantId>> and <<remotePrincipalId>> until we get clarity on using the Platform Key Vault for our development in CARML. Updated the PR description as well.
- Read me files are still going to be updated to remove the hardcoded values (this is in the parameter usage sections).
- I think we should keep the Environment Variable (Platform_Key_Vault) for now in our workflows and will remove under #470 if that was the agreement. This has no impact to existing functionality, but I do not want to bring other changes than the actual roll out of tokens here.

ahmadabdalla commented:
@eriqua @MrMCake 
- `<<resourceGroupName>>` and `<<namePrefix>>` have been rolled back. 
- `namePrefix` also removed from local token store (settings). So it doesn't get pulled by the Local-Test Script when searching for the sxx prefix
- Local Test script also removed support to replace resourceGroupName. Reason. The last stage of swapping back tokens will pickup 'validation-rg' and replace it with `<<resourceGroupName>>`

MrMCake commented on docs/wiki/GettingStarted.md:
```diff
@@ -181,7 +181,7 @@ Last but not least, instead of fetching your own copy of the repository you can
 If you are forking or cloning the repository, you can use 'tokens' inside your parameter files. Tokens allow you to test deploying modules in your own environment (i.e. using tokens for your naming conventions), or apply other customizations to your resources (i.e. using your own subscription ID inside a Resource ID string). See details in the [Parameter File Tokens Design](./ParameterFileTokens. ...
 
-The repository contains a [Settings.json](https://github.com/Azure/ResourceModules/blob/main/settings.json) that has a default token called `namePrefix`. You can use a custom value here and then start adding the tokens in your parameter files. The tokens default format is `<<tokenName>>` (i.e. `<<namePrefix>>`). Here is an example token:
+The repository contains a [Settings.json](https://github.com/Azure/ResourceModules/blob/main/settings.json) that has a default token called `namePrefix`. You can use a custom value here and then start adding the tokens in your parameter files. The tokens default format is `<<tokenName>>` (i.e. `sxx`). Here is an example token:
```
I'm not sure I get the example. Would the example not be something like `<<defaultprefix>>`?

ahmadabdalla replied:
I've made it clearer in this section. Please check

MrMCake commented on docs/wiki/GettingStarted.md:
```diff
@@ -212,7 +212,7 @@ Note: There are default tokens that can be enabled on any resource that leverage
 - `<<deploymentSpId>>`: Will point to the Service Principal ID used for deployments.
 - `<<platformKeyVault>>`: Will point to the Platform Azure Key Vault (Optional if enabled as per guidelines)
-- `<<resourceGroupName>>`: Will point to the Azure Resource Group where the resources are being deployed to. (This isn't defined in the secrets section but is injected at runtime)
+- `validation-rg`: Will point to the Azure Resource Group where the resources are being deployed to. (This isn't defined in the secrets section but is injected at runtime)
```
I think this happened due to search & replace

ahmadabdalla replied:
good pickup. fixed

MrMCake commented on docs/wiki/TestingUsage.md:
```diff
@@ -82,8 +82,8 @@ The following scenarios are common to when to use a token value in the parameter
-        "value": "/subscriptions/<<subscriptionId>>/resourceGroups/<<resourceGroupName>>/providers/Microsoft.Network/virtualNetworks/adp-<<namePrefix>>-az-vnet-x-peer01"
-        //<<namePrefix>> is a Custom Local Token
+        "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualNetworks/adp-sxx-az-vnet-x-peer01"
+        //sxx is a Custom Local Token
```
I guess the comment is now a bit redundant ;)

ahmadabdalla replied:
fixed
